### ISSUE TITLE
Pending state updates may be confusing

### ISSUE THREAD

**Issue by spicyj** [facebook/react#122]
If you have

```
componentWillReceiveProps: function() {
    this.setState({x: this.state.x + 1});
    this.setState({x: this.state.x + 1});
}
```

then `x` will only be incremented once because the new state will just be stored in `_pendingState` until `receiveProps` finishes. After my refactoring in #115, this will never increment twice, which is probably confusing.

Maybe this isn't a bug. We can fix this behavior by instead updating this.state immediately (but still updating the UI later). Right now I can't think of any situations where having this.state out of sync with the UI is a problem.

@vjeux suggested I open an issue, so here I am. Probably easiest for me to fix while updating #115 though.

**Comment by aldendaniels** [comment]
DISCLAIMER: I'm new to React.

If I understand correctly, React's `setState()` method's only advantage over directly modifying component state and calling `forceUpdate()` is that `setState()` lets the developer avoid unnecessary re-renders by overloading `shouldComponentUpdate()`.

For `shouldComponentUpdate()` to be effective, however, we need to use immutable data structures (e.g. ClosureScript + OM, React Immutability Helpers, ImmutableJS, etc.). Without immutability, `shouldComponentUpdate()` can't detect changes.

In fact, we often CAN'T use `setState()` when using mutable data. You can't remove an item from a list using `setState()` unless you've shallow cloned the whole list (think immutability).

**Takeaways**

 1. If you're using immutable data structures, then use `setState()` and enjoy the confidence that `this.state` and `this.refs` are always in sync.

 2. If you're using mutable data structures, modify `this.state` directly and call `forceUpdate()`. You'll have to live with `this.state` and `this.refs` being out of sync. On the bright side, you'll get to write VanillaJS update code and won't run into the confusion detailed in this issue.

Again - I'm new to React. Please clarify if I've got something wrong.

**Comment by leebyron** [comment]
@aldendaniels I think you've mixed two concepts.

You're 100% correct about `PureRenderMixin`'s implementation of `shouldComponentUpdate()`. It is only safe if your props and state use immutable values (including string, number, bool).

However independent of any `shouldComponentUpdate` implementation, `setState` is still perfectly safe to use, even if you're using mutable data in your props and state. 

In fact, you should **never** mutate `this.state` directly, as that *is* considered immutable.

**Comment by aldendaniels** [comment]
Hi @leebyron - thanks for reaching out.

> you should never mutate this.state directly, as that is considered immutable

I'm a little dubious: from the React docs - "React lets you use whatever style of data management you want, including mutation". I also see that in Facebook's official [TodoMVC Flux Example](https://github.com/facebook/flux/blob/master/examples/flux-todomvc/js/stores/TodoStore.js#L65), state is not being treated as immutable.

Using immutable state in JavaScript without using a 3rd party library like those listed in my previous post quickly becomes slow and tedious when dealing with large, nested data sets. 

If mutable state really is anti-pattern, then I'm all ears - but I'm also concerned by the invasiveness of the requirement.

**Comment by sebmarkbage** [comment]
There is a difference between `this.state.foo = mutation;` and `this.state.nestedObject.foo = mutation;`. The later is still accepted. The former is a bit of a gray area but is considered an anti-pattern since that object is considered "owned" by React.

**Comment by leebyron** [comment]
Yes, to clarify:

This is okay:

```js
this.state.myArray.push(newValue);
this.forceUpdate();
```

This is *not* okay:

```js
this.state.myArray = newArrayValue;
this.forceUpdate();
```

**Comment by leebyron** [comment]
But this is preferable:

```js
var pushedArray = this.state.myArray;
pushedArray.push(newValue);
this.setState({ myArray: pushedArray });
```

**Comment by dustingetz** [comment]
Take it to the mailing list please guys

**Comment by SystemParadox** [comment]
This should definitely be changed to update `this.state` immediately. The current behaviour is not only confusing- it forces us to pass the new state around to achieve sensible results.

**Comment by markmarijnissen** [comment]
Hey everybody,

Check out Meteor's implementation of their Tracker. 

They queue all state changes and flush on a requestAnimationFrame, unless you explicitly call Tracker.flush() to proces all pending changes.

if you do
```
bob_bank_account -= 10;
alice_bank_account += 10;
```

you will never end up with an invalid state (i.e. 10$ appearing or dissapearing out of nowhere). Either all changes are pending, or they are all resolved (using Tracker.flush, or at the end of the event cycle).

So, to summarize:
* `setState` should read as `enqueueStateChange`
* Add a method to process queue if you want to force the state changes immediatly (for example when you need in-between results (updated state) for you calculations)

**Comment by SystemParadox** [comment]
@markmarijnissen, your example works the same in react:

```
this.setState({ bob_account: this.state.bob_account - 10 });
this.setState({ alice_account: this.state.alice_account + 10 });
```

This specific example will work correctly. BUT, this is not advised because react will currently give you broken results if there are any *other* state changes in between time.

The problem is that reading from `this.state` doesn't give you the result you expect whilst there are pending updates. Therefore, this example is broken:

```
this.setState({ bob_account: this.state.bob_account + 10 });
this.setState({ bob_account: this.state.bob_account + 10 });
// when state resolves, bob_account is only +10, not +20 like it should have been
```

> setState should read as enqueueStateChange

No. Whilst more accurate to the current behaviour, this is not useful. We want `this.state` to be changed to the pending state.

**Comment by syranide** [comment]
@markmarijnissen @SystemParadox, `setState` can take a function as of React 0.13 and solves the problem of pending state.

**Comment by SystemParadox** [comment]
@syranide, that doesn't solve @azoerb's example, which is 99% of the problem.

I must say I am very disappointed with this change to setState in React 0.13. It doesn't solve the problem and just further solidifies the current behaviour.

**Comment by sebmarkbage** [comment]
@azoerb's checkAlarm example can be rewritten as:

```js
checkAlarm: function() {
  this.setState(state => ({
    alarmSet: state.alarmTime > 0 && state.elapsedTime < state.alarmTime
  }));
}
```

One of the most important performance improvements that React introduced was a way to avoid read/write thrash through batching using a declarative update model that doesn't depend on the order of execution. This solidifies part of the behavior that made React fast enough to begin with.

The mental model should be that state is part of the return value. You can introduce mutable data structures that allow you to mutate things as you go, but non-local mutable state is very difficult to reason about at scale so we would recommend that you use the functional approach.

This scheduling work is also part of what makes React performant and how we can make it even more performant.

**Comment by SystemParadox** [comment]
Oh I see. Thank you for the example.

I am aware that React batches rendering (this of course is why we have pending state at all), but I don't understand why the component API forces us to work with this all the time.

As @spicyj said at the beginning, why not update this.state immediately and leave the UI to update with its batching mechanism later? Anything that would be affected by this belongs in `componentDidUpdate` anyway.

**Comment by sebmarkbage** [comment]
That would only be possible with a synchronous flushing mechanism like @markmarijnissen proposed.

Take the check alarm example. It is safe for me to move elapsedTime to props like this:

```js
checkAlarm: function() {
  this.setState((state, props) => ({
    alarmSet: state.alarmTime > 0 && props.elapsedTime < state.alarmTime
  }));
}
```

However, it is not safe in the case `state` is synchronously updated, but `props` are not. E.g. this is not safe even if state is up-to-date:

```js
this.setState({
  alarmSet: this.state.alarmTime > 0 && this.props.elapsedTime < this.state.alarmTime
});
```

You would have to introduce a synchronous `this.flush();` to be able to read from props synchronously.

Now if you have two of these in the same subtree, you will have one unnecessary rerender of the entire tree. If you have 15 of these updates in the same subtree you have 14 unnecessary rerenders.

**Comment by dubrowgn** [comment]
I think the part of this that makes it inherently confusing is that there are many consumers of state, not just render. Flushing state mutations on frame animation effectively means state has to know about render, even though render is supposed to be a pure function, taking state as one of its inputs. Render's output always represents a snapshot in time anyway, so mutating state at frame animation or all along the way with each state update shouldn't affect render at all.

As it stands, state mutation and render are tied together, which isn't immediately obvious to anyone who hasn't read the docs. I would also argue that it's much less useful to have them tied together.

**Comment by SystemParadox** [comment]
For myself and anyone else who keeps bumping into this...

When the state change is asynchronous, like this:

```
getData: function () {
    var self = this;
    var params = {
        foo: this.state.foo,
        bar: this.state.bar,
    };
    getSomeDataAsynchronously(params, function (data) {
        self.setState({ data: data });
    });
},

handleFooChange: function (value) {
    this.setState({ foo: value });
    this.getData();
},

handleBarChange: function (value) {
    this.setState({ bar: value });
    this.getData();
},
```

In this case, reading the state in `getData` ends up with the previous state. The correct solution is to use the **SECOND** callback argument to `setState`, to run the update after you've finished changing the state:

```
handleFooChange: function (value) {
    this.setState({ foo: value }, this.getData);
},
```

Of course, this waits for a re-render, which isn't really what we wanted here. What we wanted to do was change the state *and* fire off the async call in the same turn, but React doesn't let us do this.

I would love to know if there is a better solution.

It also doesn't make sense to me that the change handler functions have to deal with this - why can't `getData` just use the pending state?

I still don't understand the rationale for this situation. @sebmarkbage suggested something about props but I don't see how that is an issue (perhaps I am missing the context, a complete example might help). If anything it's possibly another example of the same confusing behaviour.

**Comment by jimfb** [comment]
@SystemParadox 

```
getData: function () {
    var self = this;
    getSomeDataAsynchronously(params, function (data) {
        self.setState({ data: data });
    });
},

handleFooChange: function (value) {
    this.setState({ foo: value });
    var params = {
        foo: value,
        bar: this.state.bar,
    };
    this.getData(params);
},

handleBarChange: function (value) {
    this.setState({ bar: value });
    var params = {
        foo: this.state.foo,
        bar: value,
    };
    this.getData(params);
},
```

Does that do what you want?

I agree it's a slightly ugly hack/workaround, but assuming I'm understanding your question correctly, that should do what you want.

**Comment by spicyj** [comment]
> What we wanted to do was change the state and fire off the async call in the same turn

Is there a reason this is effectively different from doing them separately?

**Comment by SystemParadox** [comment]
@jimfb, that works, but duplicating the parameter handling is worse than having to pass the 'correct' state object each time.

@spicyj, arguably it doesn't make a lot of practical difference, but it's harder to reason about and it separates the async call from the handler that caused it, which can hinder debugging. If there's no reason to complicate things by separating them with an extra turn, why are we doing it?

**Comment by spicyj** [comment]
Closing this out because we're not going to change anything, at least with our current component syntax. The recommended way to do transactional updates is:

```
    this.setState((state) => ({x: state.x + 1}));
    this.setState((state) => ({x: state.x + 1}));
```

**Comment by arcanis** [comment]
Transactional updates are a thing, but what if we only need to read the actual value? For example, in the following pseudo-component I listen on some sort of global event listener to toggle a "selectable" internal flag. However, I can't trust this field to be actually set in the `onClickHandler` call, which means that as far as I know, I either have to manually keep track of the current state (which seems like a quite nasty way to solve this issue), or wrap the code inside a dummy `setState` call that wouldn't do anything.

```
class Foo extends Component {

    static contextTypes = {
        globalEmitter : React.PropTypes.object
    };

    componentDidMount() {
        this.context.globalEmitter.on('selectstart', this.onSelectStart);
        this.context.globalEmitter.on('selectstop', this.onSelectStop);
    }

    componentWillUnmount() {
        this.context.globalEmitter.off('selectstart', this.onSelectStart);
        this.context.globalEmitter.off('selectstop', this.onSelectStop);
    }

    onClickHandler = () => {
        if (!this.state.selectable)
            return ;
        this.props.onSelected();
    };

    onSelectStart = () => {
        this.setState({ selectable: true });
    };

    onSelectStop = () => {
        this.setState({ selectable: false });
    };

}
```

**Comment by azoerb** [comment]
Maël,

In that case, I'd just add a non-state class variable:

```
class Foo extends Component {
  selectable: false,  // New variable here
  ...
  onClickHandler = () => {
    if (!this.selectable) return;
    ...
  }

  onSelectState = () => {
    this.selectable = true;
  }
  ...
}
```

State should really only be used for rendering purposes, so if you need a
variable that doesn't have anything to do with the render method, use a
class variable instead.

**Comment by azoerb** [comment]
Maël,

In that case, I'd just add a non-state class variable:

```
class Foo extends Component {
  selectable: false,  // New variable here
  ...
  onClickHandler = () => {
    if (!this.selectable) return;
    ...
  }

  onSelectState = () => {
    this.selectable = true;
  }
  ...
}
```

State should really only be used for rendering purposes, so if you need a variable that doesn't have anything to do with the render method, use a class variable instead.

**Comment by giorgim** [comment]
@spicyj Indeed state management in react is the single place, which got me confused in the first place. Mainly probably due its asynchronous nature, etc. I think from the perspective of the consumer of react, the API you have suggested for managing state, is not the most intuitive and straightforward to understand (despite all the technical benefits it may have for the library itself).

**Comment by gaearon** [comment]
Wrote some thoughts about why React works this way: https://github.com/facebook/react/issues/11527#issuecomment-360199710
